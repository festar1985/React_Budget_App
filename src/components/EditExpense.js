import React from "react";

const EditExpense = (props) => <p>Test Edit on id {props.match.params.id}</p>;

export default EditExpense;
