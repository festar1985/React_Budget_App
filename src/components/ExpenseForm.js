import "react-dates/initialize";
import React from "react";
import { useState } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const ExpenseForm = (props) => {
  const [state, setState] = useState({
    description: props.expense ? props.expense.description : "",
    note: props.expense ? props.expense.note : "",
    amount: props.expense ? (props.expense.amount / 100).toString() : "",
    calendarFocused: false,
  });

  const [dateState, setDateState] = useState({
    createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
  });

  const [errorState, setErrorState] = useState({
    error: false,
  });

  const onDateChange = (date) =>
    date ? setDateState({ createdAt: date }) : dateState.createdAt;

  const onFocusChange = ({ focused }) => {
    setState({
      ...state,
      calendarFocused: focused,
    });
  };

  const onSubmit = (e) => {
    e.preventDefault();
    if (!state.description || !state.amount) {
      setErrorState({ error: true });
    } else {
      setErrorState({ error: false });
      props.onSubmit({
        ...state,
        amount: parseFloat(state.amount, 10) * 100,
        createdAt: dateState.createdAt.valueOf(),
      });
    }
  };

  return (
    <div>
      {errorState.error && <p>Please fill both description and ammound</p>}
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Description"
          autoFocus
          value={state.description}
          onChange={(e) =>
            setState({
              ...state,
              description: e.target.value,
            })
          }
        ></input>
        <input
          type="text"
          placeholder="Amount"
          value={state.amount}
          onChange={(e) => {
            let amount = e.target.value;

            if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/gm)) {
              setState({
                ...state,
                amount: amount,
              });
            }
          }}
        ></input>
        <SingleDatePicker
          date={dateState.createdAt}
          onDateChange={onDateChange}
          focused={state.calendarFocused}
          onFocusChange={onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note (optional)"
          value={state.note}
          onChange={(e) =>
            setState({
              ...state,
              note: e.target.value,
            })
          }
        ></textarea>
        <button>Add Expense</button>
      </form>
    </div>
  );
};

export default ExpenseForm;
