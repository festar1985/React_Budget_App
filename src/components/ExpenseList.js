import React from "react";
import { connect } from "react-redux";
import ExpensesListItem from "./ExpenseListItem";
import getVisibleExpenses from "../redux/selectors/expenses";

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ? (
      <p>There are no expenses to show</p>
    ) : (
      props.expenses.map((element) => {
        return <ExpensesListItem key={element.id} {...element} />;
      })
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
