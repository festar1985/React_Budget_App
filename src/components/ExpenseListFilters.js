import React from "react";
import { connect } from "react-redux";
import {
  filterText,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../redux/actions/filters";
import { useState } from "react";
import { DateRangePicker } from "react-dates";

const ExpenseListFilters = (props) => {
  const [state, setState] = useState({
    calendarFocus: null,
  });

  const onDatesChange = ({ startDate, endDate }) => {
    props.dispatch(setStartDate(startDate));
    props.dispatch(setEndDate(endDate));
  };

  const onFocusChange = (calendarFocus) => {
    setState(() => ({ calendarFocus }));
  };

  return (
    <div>
      <input
        type="text"
        value={props.filters.text}
        onChange={(event) => {
          props.dispatch(filterText(event.target.value));
        }}
      />
      <select
        value={props.filters.sortBy}
        onChange={(event) => {
          if (event.target.value === "date") {
            props.dispatch(sortByDate(event.target.value));
          } else if (event.target.value === "amount") {
            props.dispatch(sortByAmount(event.target.value));
          }
        }}
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={props.filters.startDate} // momentPropTypes.momentObj or null,
        endDate={props.filters.endDate} // momentPropTypes.momentObj or null,
        onDatesChange={onDatesChange} // PropTypes.func.isRequired,
        focusedInput={state.calendarFocus} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
        onFocusChange={onFocusChange} // PropTypes.func.isRequired,
        numberOfMonths={1}
        isOutsideRange={() => false}
        showClearDates={true}
        keepOpenOnDateSelect={true}
      />
    </div>
  );
};

const mapStateToProps = (state) => {
  return { filters: state.filters };
};

export default connect(mapStateToProps)(ExpenseListFilters);
