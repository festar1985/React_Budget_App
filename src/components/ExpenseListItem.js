import React from "react";

import { Link } from "react-router-dom";

const ExpensesListItem = ({ id, description, category, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{category}</p>
    <p>{amount}</p>
    <p>{createdAt}</p>
  </div>
);

export default ExpensesListItem;
