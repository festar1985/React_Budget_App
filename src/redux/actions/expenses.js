import { push, onValue, remove, ref, update } from "firebase/database";
import { myDatabase, databaseReference } from "../../firebase/firebase";

//Actions

export const addExpense = (expenses) => ({
  type: "ADD_EXPENSE",
  expenses,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "Description not available",
      category = "Others",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, category, note, amount, createdAt };

    push(databaseReference, expense).then((reference) => {
      dispatch(
        addExpense({
          id: reference.key,
          ...expense,
        })
      );
    });
  };
};

export const removeExpense = (id = "") => ({ type: "REMOVE_EXPENSE", id });

export const startRemoveExpense = (id) => {
  return async (dispatch) => {
    await remove(ref(myDatabase, id));
    dispatch(removeExpense(id));
  };
};

export const modifyExpense = (id, updates) => ({
  type: "MODIFY_EXPENSE",
  id,
  updates,
});

export const startModifyExpense = (id, updates) => {
  return async (dispatch) => {
    await update(ref(myDatabase, id), updates);
    dispatch(modifyExpense(id, updates));
  };
};

export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    onValue(
      databaseReference,
      (snapshot) => {
        const expenses = [];
        Object.entries(snapshot.val()).forEach(([key, val]) => {
          const expense = {
            id: key,
            ...val,
          };
          expenses.push(expense);
        });

        dispatch(setExpenses([...expenses]));
      },
      { onlyOnce: true }
    );
  };
};
