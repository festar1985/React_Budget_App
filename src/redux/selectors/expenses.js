const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((element) => {
      const startDateMatch =
        typeof startDate != "number" || element.createdAt >= startDate;
      const endDateMatch =
        typeof startDate != "number" || element.createdAt <= endDate;
      const textMatch = element.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
