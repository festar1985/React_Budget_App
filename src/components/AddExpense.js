import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../redux/actions/expenses";

export const AddExpense = (props) => (
  <div>
    <h1>Add an expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        //props.dispatch(addExpense(expense));
        props.onSubmit(expense);
        props.history.push("/");
      }}
    />
  </div>
);

const mapDispatchToProps = (dispatch) => {
  return {
    onSubmit: (expense) => dispatch(startAddExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpense);
