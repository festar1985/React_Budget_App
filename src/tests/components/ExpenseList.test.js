import React from "react";
import { ExpenseList } from "../../components/ExpenseList";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";
import { expenses } from "../fixture/expenses";

Enzyme.configure({ adapter: new Adapter() });

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render ExpenseList with empty expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
