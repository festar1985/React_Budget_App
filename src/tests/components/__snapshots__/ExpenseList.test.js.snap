// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty expenses 1`] = `
<div>
  <p>
    There are no expenses to show
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpensesListItem
    amount={30}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpensesListItem
    amount={20}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpensesListItem
    amount={40}
    createdAt={345600000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
